src/lib/QueryType.hs

```diff
@@ -358,18 +359,30 @@ instance HasType Atom where
                                      (Var v') -> return v'
                                      _ -> error "!?"
                                    instantiateDepPairTy t (ProjectElt (0 NE.:| is) v')
-        -- Newtypes
-        TC (Fin _) | i == 0 -> return IdxRepTy
-        StaticRecordTy types | i == 0 -> return $ ProdTy $ toList types
-        TypeCon _ defName params | i == 0 -> do
-          def <- lookupDataDef defName
-          [DataConDef _ repTy _] <- instantiateDataDef def params
-          return repTy
-        RecordTy _ -> throw CompilerErr "Can't project partially-known records"
+        _ | isNewtype ty && i == 0 -> projectNewtype ty
         Var _ -> throw CompilerErr $ "Tried to project value of unreduced type " <> pprint ty
         _ -> throw TypeErr $
               "Only single-member ADTs and record types can be projected. Got " <> pprint ty
 
+isNewtype :: Type n -> Bool
+isNewtype ty = case ty of
+  TC _          -> True
+  TypeCon _ _ _ -> True
+  RecordTy _    -> True
```

Comment:
Variant type is missing? It might be working out because you only use this in projection handling and you can't project from variants, but it's a bug waiting to happen